udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala

```diff
@@ -0,0 +1,128 @@
+    val className = constPool.getFieldrefClassName(constPoolIndex)
+    classPool.getCtClass(className).getField(fieldName, descriptor)
+  }
+
+  def lookupBehavior(constPoolIndex: Int): CtBehavior = {
+    if (constPool.getTag(constPoolIndex) != ConstPool.CONST_Methodref) {
+      throw new SparkException("Unexpected index for method reference")
+    }
+    val methodName = constPool.getMethodrefName(constPoolIndex)
+    val descriptor = constPool.getMethodrefType(constPoolIndex)
+    val className = constPool.getMethodrefClassName(constPoolIndex)
+    val params = Descriptor.getParameterTypes(descriptor, classPool)
+    if (constPool.isConstructor(className, constPoolIndex) == 0) {
+      classPool.getCtClass(className).getDeclaredMethod(methodName, params)
+    } else {
+      classPool.getCtClass(className).getDeclaredConstructor(params)
+    }
+  }
+
+  def lookupClassName(constPoolIndex: Int): String = {
+    if (constPool.getTag(constPoolIndex) != ConstPool.CONST_Class) {
+      throw new SparkException("Unexpected index for class")
+    }
+    constPool.getClassInfo(constPoolIndex)
+  }
+
+  // Get the CtClass object for the class that capture the lambda.
+  private val ctClass = {
+    val name = serializedLambda.getCapturingClass.replace('/', '.')
+    val loader = Thread.currentThread().getContextClassLoader
+    // scalastyle:off classforname
+    val classForName = Class.forName(name, true, loader)
+    // scalastyle:on classforname
+    classPool.insertClassPath(new ClassClassPath(classForName))
+    classPool.getCtClass(name)
+  }
+
+  // Get the CtMethod object for the method that implements the lambda body.
+  private val ctMethod = {
+    val lambdaImplName = serializedLambda.getImplMethodName
+    ctClass.getDeclaredMethod(lambdaImplName.stripSuffix("$adapted"))
+  }
+
+  private val methodInfo = ctMethod.getMethodInfo
+
+  val constPool = methodInfo.getConstPool
+
+  private val codeAttribute = methodInfo.getCodeAttribute
+
+  lazy val codeIterator: CodeIterator = codeAttribute.iterator
+
+  lazy val parameters: Array[CtClass] = ctMethod.getParameterTypes
+
+  lazy val ret: CtClass = ctMethod.getReturnType
+
+  lazy val maxLocals: Int = codeAttribute.getMaxLocals
+}
+
+object LambdaReflection {
+  def apply(function: AnyRef): LambdaReflection = {
+    // writeReplace is supposed to return an object of SerializedLambda from
+    // the function class (See
+    // https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/SerializedLambda.html).
+    // With the object of SerializedLambda, we can get our hands on the class
+    // and the method that implement the lambda body.
+    val functionClass = function.getClass
+    val writeReplace = functionClass.getDeclaredMethod("writeReplace")
+    writeReplace.setAccessible(true)
+    val serializedLambda = writeReplace.invoke(function)
+      .asInstanceOf[SerializedLambda]
+
+    // ClassPool: javassist TODO: watch for this object getting huge
```

Comment:
> ClassPool objects hold all the CtClasses that have been created so that the consistency among modified classes can be guaranteed. Thus if a large number of CtClasses are processed, the ClassPool will consume a huge amount of memory. To avoid this, a ClassPool object should be recreated, for example, every hundred classes processed. Note that getDefault() is a singleton factory. Otherwise, detach() in CtClass should be used to avoid huge memory consumption.

It seems we should call `detach` on the `CtClass` when done. Any reason not to do that @seanprime7?

Reply:
I will check and let you know here.

Reply:
I think the following diff should be simpler to manage than calling `detach` on every `CtClass` that is created directly or indirectly by `LambdaReflection`.

The following change will allow each reflection to have its own class pool which will be garbage collected along with its `CtClass` objects when the reflection is garbage collected.

```
diff --git a/udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala b/udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala
index 45305a06..74cc3dfe 100644
--- a/udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala
+++ b/udf-compiler/src/main/scala/com/nvidia/spark/udf/LambdaReflection.scala
@@ -118,9 +118,7 @@ object LambdaReflection {
     val serializedLambda = writeReplace.invoke(function)
       .asInstanceOf[SerializedLambda]

-    // ClassPool: javassist TODO: watch for this object getting huge
-    // http://www.javassist.org/html/javassist/ClassPool.html
-    val classPool = ClassPool.getDefault
+    val classPool = new ClassPool(true)
     LambdaReflection(classPool, serializedLambda)
   }
 }
```

Reply:
Sounds good to me. The one concern is that they say to do this "occasionally", so perhaps this instantiation cost is not cheap?

Reply:
I have made this change locally and ran the tests in `OpcodeSuite` and I don't see much difference in perf.
I think they say "occasionally" because

- a new `ClassPool` object has no `CtClass` object loaded
- when a new `CtClass` object is created in a `ClassPool` object, the corresponding `.class` file needs to be read, which may be slow as it is a disk I/O.

This may be an issue if we are "replacing" the `ClassPool` object used by the entire module.  However, with the code diff above, we will allow each `LambdaReflection` object to load a class that it needs only once and avoid redundant disk I/O without adding additional burden on the default `ClassPool` object.

Anyway, if the instantiation cost turns out to be too high, we can explore other approaches.

Reply:
Sounds good. We should add some timings to this code. We use nvtx traces in code run in the executor, I am not sure we have that in the driver. At least a simple start/end with a log line showing time to see how much time the compiler takes from start to finish.

Reply:
@wjxiz1992 fyi.

Reply:
@abellina I just bring Sean's original history back. I'll address those when he's fine with my approach. Thanks for the info!

Reply:
Updated.